app/controllers/concerns/checkout_update.rb

```diff
@@ -19,6 +20,7 @@ def update_payment
 
     def finalize_order
       current_order.update(status: :in_delivery)
+      current_order.coupon.update(active: false) if current_order.coupon.present?
```

Comment:
This is not refactoring. You just moved line from one action to another.
Maybe it would be better to encapsulate all these methods + sending mail to some service?

Reply:
It will be harder to read and understand code. Just a two-three lines are not worth for creating a new service.

Reply:
I totally disagree with you.
You are updating 2 objects and sends mails from controller actions. Your controller has to be as thin as possible!